### PR TITLE
add network-level subgraph

### PR DESCRIPTION
**Pull request by dmelotik**
Deliverables:

- A subgraph for all mainnet networks supported by thegraph's hosted service
- A standard schema for network metrics (evm, cosmos, arweave, near, cronos)

Deployments: https://okgraph.xyz/?q=messari%2Fnetwork-[NETWORK_NAME]
Cronos deployment: https://graph.cronoslabs.com/subgraphs/name/messari/network-cronos

### CODE REVIEW

File: subgraphs/layer-one/abis/Gravity.json [hunk 1]
```diff
@@ -0,0 +1,106 @@
+[
```
Comment (this-username-is-taken) [inline subgraphs/layer-one/abis/Gravity.json:1]: delete

Comment (this-username-is-taken) [inline subgraphs/layer-one/abis/Gravity.json:1]: delete

File: subgraphs/layer-one/Makefile [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# run make to do a full deployment
```
Comment (this-username-is-taken) [inline subgraphs/layer-one/Makefile:1]: you should adopt @steegecs's deployment flow

Comment (this-username-is-taken) [inline subgraphs/layer-one/Makefile:1]: you should adopt @steegecs's deployment flow

File: subgraphs/layer-one/.prettierignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+build
+generated
```
Comment (this-username-is-taken) [inline subgraphs/layer-one/.prettierignore:2]: add new line at EOF

Comment (this-username-is-taken) [inline subgraphs/layer-one/.prettierignore:2]: add new line at EOF

Comment (this-username-is-taken) [reply subgraphs/layer-one/.prettierignore:2]: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

Comment (this-username-is-taken) [reply subgraphs/layer-one/.prettierignore:2]: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

**Comment by this-username-is-taken** [comment]
Sources to use as a reference for data to include: https://docs.coinmetrics.io/info/metrics

**Comment by this-username-is-taken** [comment]
Can we rename L1/layer-one to `network`? It's a more appropriate name as some of the chains are not L1, also to align with our existing subgraphs terminology.

File: subgraphs/layer-one/abis/Gravity.json (hunk 1 above)
Comment (dmelotik) [reply subgraphs/layer-one/abis/Gravity.json:1]: i have to have one abi in the manifest, but i was able to remove everything in between the brackets

Comment (this-username-is-taken) [reply subgraphs/layer-one/abis/Gravity.json]: ah, gotcha. Maybe name it something that makes more sense, like `Placeholder`

File: schema-generic.graphql [hunk 4]
```diff
@@ -6,16 +6,21 @@ enum Network {
...
   BSC # aka BNB Chain
   CELO
+  CHAPEL
+  CLOVER
   MAINNET # Ethereum Mainnet
   FANTOM
   FUSE
   MOONBEAM
   MOONRIVER
+  MUMBAI
```
Comment (this-username-is-taken) [inline schema-generic.graphql:19]: isn't mumbai a testnet?

File: schema-generic.graphql [hunk 5]
```diff
@@ -6,16 +6,21 @@ enum Network {
   ARBITRUM_ONE
   AURORA
   AVALANCHE
+  BOBA
   BSC # aka BNB Chain
   CELO
+  CHAPEL
```
Comment (this-username-is-taken) [inline schema-generic.graphql:12]: Do you know what Chapel, Clover, and POA are? I've never heard of them. I thought xDAI used to be POA

File: subgraphs/network/schema.graphql [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+###############################
+
+type DailySnapshot @entity {
+  " { number of days since UNIX epoch time } "
+  id: ID!
+
+  " ID of the network this snapshot belongs to "
+  network: Network!
+
+  " Number of cumulative ununique authors "
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:91]: typo

File: subgraphs/network/schema.graphql [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  " The version of the subgraph "
+  subgraphVersion: String!
+
+  " The version of the methodology "
+  methodologyVersion: String!
+
+  #### Quantitative Data ####
+
+  " Number of cumulative unique authors "
+  cumulativeUniqueAuthors: Int!
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:48]: what's an `author`?

File: subgraphs/network/schema.graphql [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+  " All of the blocks created during this day "
+  blocks: [Block!]! @derivedFrom(field: "hour")
+}
+
+##########################
+#### Block-level Data ####
+##########################
+
+type Block @entity {
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:236]: are we able to get transactions per block?

Comment (dmelotik) [reply subgraphs/network/schema.graphql:236]: not on evm or near, will check on the other two

Comment (this-username-is-taken) [reply subgraphs/network/schema.graphql:236]: hmm, okay. That's unfortunate :(

File: subgraphs/layer-one/abis/Gravity.json (hunk 1 above)
Comment (dmelotik) [reply subgraphs/layer-one/abis/Gravity.json]: good idea - done

File: schema-generic.graphql (hunk 4 above)
Comment (dmelotik) [reply schema-generic.graphql:19]: yes you are right, polygon testnet. removed

File: schema-generic.graphql (hunk 5 above)
Comment (dmelotik) [reply schema-generic.graphql:12]: Chapel is BSC testnet - removed

Clover seems to be a polkadot parachain - https://parachains.info/details/clover_finance

and yeah i found out poa finished merging into gnosis on may 31 2022 and POA token is not supposed to have value anymore. It is now a decentralized testnet essentially

**Comment by this-username-is-taken** [comment]
Would you be able to integrate these improvements in as well?
https://medium.com/edge-node-engineering/two-simple-subgraph-performance-improvements-a76c6b3e7eac

File: subgraphs/network/schema.graphql [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+  cumulativeDifficulty: BigInt!
+
+  " Total amount of native asset created "
+  cumulativeRewards: BigInt
+
+  " A mapping to every block that this author has generated "
+  blocks: [Block!]! @derivedFrom(field: "author")
+}
+
+enum IntervalType {
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:301]: is this needed?

File: subgraphs/network/schema.graphql [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+  reward: BigInt
+
+  " Hash of the previous block (TODO: is this necessary??) "
+  parentHash: String!
+
+  " Burnt fees in native asset (baseFeePerGas * gasUsed) (note, post London upgrade) "
+  burntFees: BigInt!
+
+  " The day this block was created (in days since UNIX epoch time) "
+  day: DailySnapshot!
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:277]: what's this used for?

File: subgraphs/network/schema.graphql [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+  cumulativeDifficulty: BigInt!
+
+  " Total gas limit provided from all transactions this block "
+  gasLimit: BigInt!
+
+  " Total gas used in this block "
+  gasUsed: BigInt!
+
+  " Percentage of gas filled in this block (0-100)"
+  gasFilledPercentage: BigDecimal!
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:265]: shouldn't this be called `blockUtilization` or `blockUsage`?
https://medium.com/@ASvanevik/why-all-these-empty-ethereum-blocks-666acbbf002

File: subgraphs/network/schema.graphql (hunk 9 above)
Comment (dmelotik) [reply subgraphs/network/schema.graphql:301]: I mean i am not sure how useful that would be..  sounds like it should be removed?

File: subgraphs/network/schema.graphql (hunk 10 above)
Comment (dmelotik) [reply subgraphs/network/schema.graphql:277]: Keep track of the Blocks in a given day, but that could  also be easily aggregated from the block number metrics in the daily snapshot. removing

File: subgraphs/network/schema.graphql (hunk 9 above)
Comment (dmelotik) [reply subgraphs/network/schema.graphql:301]: removing

File: subgraphs/network/schema.graphql [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+  cumulativeDifficulty: BigInt!
+
+  " Difficulty during this day "
+  dailyDifficulty: BigInt!
+
+  " Average difficulty for this day "
+  dailyMeanDifficulty: BigDecimal!
+
+  " Sum of all gas used during this day "
+  dailyCumulativeGasUsed: BigInt!
```
Comment (this-username-is-taken) [inline subgraphs/network/schema.graphql:124]: is it possible to calculate cumulative fees as well?

File: subgraphs/network/schema.graphql [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+
+  " Number of transactions during this day "
+  hourlyTransactionCount: Int
+}
+
+##########################
+#### Block-level Data ####
+##########################
+
+type Block @entity(immutable: true) {
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/network/schema.graphql:301]: done - only possible on some networks however

File: subgraphs/network/README.md [hunk 14]
```diff
@@ -0,0 +1,194 @@
+# EVM L1 Subgraph
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/network/README.md:1]: thanks!

File: subgraphs/network/schema.graphql [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+  " Total gas used in this block "
+  gasUsed: BigInt!
+
+  " Percentage of gas filled in this block (e.g. gasUsed / gasLimit) "
+  blockUtilization: BigDecimal!
+
+  " The price of gas in this block in network native gas units "
+  gasPrice: BigInt
+
+  " Hash of the previous block (TODO: is this necessary??) "
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/network/schema.graphql:335]: removed

File: subgraphs/network/schema.graphql (hunk 12 above)
Comment (dmelotik) [reply subgraphs/network/schema.graphql:149]: not possible on EVM b/c we don't have gasPrice: https://ethereum.org/ig/developers/docs/gas/#pre-london

from my understanding the burntFees in near are the transaction fees: https://near.org/blog/near-protocol-economics/

cosmos also needs gasPrice which we don't have: https://docs.cosmos.network/master/basics/gas-fees.html#introduction-to-gas-and-fees

Arweave: TODO

Comment (this-username-is-taken) [reply subgraphs/network/schema.graphql:149]: ah, no problem

File: subgraphs/network/README.md [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
```
Comment (this-username-is-taken) [inline subgraphs/network/README.md:1]: Will be good to rename this to `Network Subgraph` as now it's grown past EVM (and L1) 😉

File: subgraphs/network/README.md [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+
+## Quickstart
+
+### Deployment
+
+Once everything is setup properly deploying is very easy.
+
+```bash
+# This example will deploy all evm networks to the hosted service under "dmelotik" in deploymentConfigurations.json
+npm run deploy --SUBGRAPH=network --POTOCOL=evm --LOCATION=dmelotik
```
Comment (this-username-is-taken) [inline subgraphs/network/README.md:13]: `PROTOCOL`

File: schema-layer-one.graphql [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+  CHUNKS
+  SUPPLY
+  TRANSACTIONS
+  BLOCKS
+  BLOCK_INTERVAL
+  GAS_PRICE
+}
+
+# Stores statistical analytics for given metrics
+type STATS @entity {
```
Comment (this-username-is-taken) [inline schema-layer-one.graphql:47]: this should be `Stats` to be consistent in naming

File: schema-layer-one.graphql [hunk 19]
```diff
@@ -0,0 +1,409 @@
+# Subgraph Schema: Network
```
Comment (this-username-is-taken) [inline schema-layer-one.graphql:1]: also rename this to `schema-network.graphql`

File: subgraphs/network/README.md [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+
+> Setting `deploy-on-merge` to `true` in [deploymentConfigurations.json](../../deployment/deploymentConfigurations.json) will run the above commands on subgraphs that have changed to messari's hosted service.
+
+## Resources and Links
+
+<details>
+<summary>See hosted service testing endpoints</summary>
+<br>
+
+- Arbitrum One: https://thegraph.com/hosted-service/subgraph/dmelotik/network-arbitrum-one
```
Comment (this-username-is-taken) [inline subgraphs/network/README.md:280]: can remove these now as they are now deployed on the official subgraphs

**Comment by this-username-is-taken** [comment]
Congrats on your first end-to-end subgraph with schema design! Fantastic work on this!

I added some minor comments. Feel free to address them in a separate PR. I will merge in this one so the next revision can be relatively easy to review.

**Comment by dmelotik** [comment]
> Congrats on your first end-to-end subgraph with schema design! Fantastic work on this!
> 
> I added some minor comments. Feel free to address them in a separate PR. I will merge in this one so the next revision can be relatively easy to review.

Glad I had the opportunity to build this *rocket emoji*! I will make a seperate PR for the changes

**Pull request merged**